python/arcticdb/storage_fixtures/s3.py

```diff
@@ -0,0 +1,275 @@
+            self.key = factory._create_user_get_key(bucket + "_user")
+        else:
+            self.key = factory.default_key
+
+        secure, host, port = re.match(r"(?:http(s?)://)?([^:/]+)(?::(\d+))?", factory.endpoint).groups()
+        self.arctic_uri = f"s3{secure or ''}://{host}:{self.bucket}?access={self.key.id}&secret={self.key.secret}"
+        if port:
+            self.arctic_uri += f"&port={port}"
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.factory.cleanup_bucket(self)
+
+    def create_test_cfg(self, lib_name: str) -> EnvironmentConfigsMap:
+        cfg = EnvironmentConfigsMap()
+        add_s3_library_to_env(
+            cfg,
+            lib_name=lib_name,
+            env_name=Defaults.ENV,
+            credential_name=self.key.id,
+            credential_key=self.key.secret,
+            bucket_name=self.bucket,
+            endpoint=self.factory.endpoint,
+            with_prefix=False,  # to allow s3_store_factory reuse_name to work correctly
+            is_https=self.factory.endpoint.startswith("s3s:"),
+            region=self.factory.region,
+        )
+        return cfg
+
+    def set_permission(self, *, read: bool, write: bool):
+        factory = self.factory
+        assert factory.enforcing_permissions and self.key is not factory.default_key
+        if read:
+            factory._iam_admin.put_user_policy(
+                UserName=self.key.user_name,
+                PolicyName="bucket",
+                PolicyDocument=factory._RW_POLICY if write else factory._RO_POLICY,
+            )
+        else:
+            factory._iam_admin.delete_user_policy(UserName=self.key.user_name, PolicyName="bucket")
+
+    def get_boto_bucket(self):
+        """Lazy singleton. Not thread-safe."""
+        if not self._boto_bucket:
+            self._boto_bucket = self.factory._boto("s3", self.key, api="resource").Bucket(self.bucket)
+        return self._boto_bucket
+
+    def iter_underlying_object_names(self):
+        return (obj.key for obj in self.get_boto_bucket().objects.all())
+
+    def copy_underlying_objects_to(self, destination: "S3Bucket"):
+        source_client = self.factory._boto("s3", self.key)
+        dest = destination.get_boto_bucket()
+        for key in self.iter_underlying_object_names():
+            dest.copy({"Bucket": self.bucket, "Key": key}, key, SourceClient=source_client)
+
+
+class BaseS3StorageFixtureFactory(StorageFixtureFactory):
+    """Logic and fields common to real and mock S3"""
+
+    endpoint: str
+    region: str
+    default_key: Key
+    default_bucket: Optional[str] = None
+    default_prefix: Optional[str] = None
+    clean_bucket_on_fixture_exit = True
+
+    def __str__(self):
+        return f"{type(self).__name__}[{self.default_bucket or self.endpoint}]"
+
+    @staticmethod
+    def from_environment_variables():
+        out = BaseS3StorageFixtureFactory()
+        out.endpoint = os.getenv("ARCTICDB_REAL_S3_ENDPOINT")
```

Comment:
Bit confusing here since the class is to `Logic and fields common to real and mock S3` but these env vars are specific to real storages?

Reply:
I can move this out of this class.